src/snarks/groth16/mpc_utils.tcc

```diff
@@ -44,87 +63,112 @@ srs_mpc_layer_L1<ppT> mpc_compute_linearcombination(
+                B_j_at_x_g2 = B_j_at_x_g2 + (entry.second *
+                                             lagrange.lagrange_g2[entry.first]);
+                ABC_j_at_x =
+                    ABC_j_at_x +
+                    (entry.second * lagrange.alpha_lagrange_g1[entry.first]);
+            }
+
+            Bs_g1[j] = B_j_at_x_g1;
+            Bs_g2[j] = B_j_at_x_g2;
+        }
+
+        {
+            G1 C_j_at_x = G1::zero();
+            const std::map<size_t, Fr> &C_j_lagrange =
+                qap.C_in_Lagrange_basis[j];
+            for (const auto &entry : C_j_lagrange) {
+                C_j_at_x =
+                    C_j_at_x + entry.second * lagrange.lagrange_g1[entry.first];
+            }
+
+            Cs_g1[j] = C_j_at_x;
+            ABC_j_at_x = ABC_j_at_x + C_j_at_x;
+        }
+
+        ABCs_g1[j] = ABC_j_at_x;
     }
+    libff::leave_block("computing A_i, B_i, C_i, ABC_i at x");
 
-    // Compute [ beta.A_i(x) + alpha.B_i(x) + C_i(x) ]_1
-    //
-    // For each i, get the Lagrange factors of A_i, B_i, C_i.  For
-    // each j, if A, B or C has a non-zero factor, grab the Lagrange
-    // coefficients, evaluate at [t]_1, multiply by the factor and
-    // accumulate.
-    libff::G1_vector<ppT> A_i_g1(num_variables + 1);
-    libff::G1_vector<ppT> B_i_g1(num_variables + 1);
-    libff::G2_vector<ppT> B_i_g2(num_variables + 1);
-    libff::G1_vector<ppT> ABC_i_g1(num_variables + 1);
-
-    evaluator_from_lagrange<ppT, G1> tau_eval_g1(pot.tau_powers_g1, domain);
-    evaluator_from_lagrange<ppT, G2> tau_eval_g2(pot.tau_powers_g2, domain);
-    evaluator_from_lagrange<ppT, G1> alpha_tau_eval(
-        pot.alpha_tau_powers_g1, domain);
-    evaluator_from_lagrange<ppT, G1> beta_tau_eval(
-        pot.beta_tau_powers_g1, domain);
-
-    for (size_t i = 0; i < num_variables + 1; ++i) {
-        // Compute [beta.A_i(x)], [alpha.B_i(x)] . [C_i(x)]
-        const std::map<size_t, Fr> &A_i_in_lagrange =
-            qap.A_in_Lagrange_basis[i];
-        const std::map<size_t, Fr> &B_i_in_lagrange =
-            qap.B_in_Lagrange_basis[i];
-        const std::map<size_t, Fr> &C_i_in_lagrange =
-            qap.C_in_Lagrange_basis[i];
-
-        A_i_g1[i] = tau_eval_g1.evaluate_from_lagrange_factors(A_i_in_lagrange);
-        B_i_g1[i] = tau_eval_g1.evaluate_from_lagrange_factors(B_i_in_lagrange);
-        B_i_g2[i] = tau_eval_g2.evaluate_from_lagrange_factors(B_i_in_lagrange);
-
-        G1 beta_A_at_t =
-            beta_tau_eval.evaluate_from_lagrange_factors(A_i_in_lagrange);
-        G1 alpha_B_at_t =
-            alpha_tau_eval.evaluate_from_lagrange_factors(B_i_in_lagrange);
-        G1 C_at_t = tau_eval_g1.evaluate_from_lagrange_factors(C_i_in_lagrange);
-
-        ABC_i_g1[i] = beta_A_at_t + alpha_B_at_t + C_at_t;
+    // Domain uses n-roots of unity, so
+    //      t(x)       = x^n - 1
+    //  =>  t(x) . x^i = x^(n+i) - x^i
+    libff::G1_vector<ppT> t_x_pow_i(n - 1, G1::zero());
```

Comment:
I'd move the computation of the set `{[t(x) . x^i]}, i in [0 .. n-2]` as the first step of this function as this is closer to the paper (and we do not need the LAG_x vector to compute the elements of this set).

Reply:
I've moved the computation of {[t(x) . x^i]} to the start of the function.